src/neo/Network/P2P/Payloads/NotaryRequest.cs

```diff
@@ -0,0 +1,136 @@
+        /// The transaction need Notary to collect signatures.
+        /// </summary>
+        private Transaction mainTransaction;
+
+        /// <summary>
+        /// This transaction is valid when MainTransaction failed.
+        /// </summary>
+        private Transaction fallbackTransaction;
+
+        /// <summary>
+        /// The witness of the payload. It must be one of multi-sig address of <see cref="MainTransaction"/>.
+        /// </summary>
+        private Witness witness;
+        private UInt256 hash = null;
+
+        public InventoryType InventoryType => InventoryType.Notary;
+
+        public UInt256 Hash
+        {
+            get
+            {
+                if (hash == null)
+                {
+                    hash = this.CalculateHash();
+                }
+                return hash;
+            }
+        }
+
+        public Witness[] Witnesses
+        {
+            get
+            {
+                return new Witness[] { witness };
+            }
+            set
+            {
+                witness = value[0];
+            }
+        }
+
+        public Transaction MainTransaction
+        {
+            get => mainTransaction;
+            set
+            {
+                mainTransaction = value;
+                hash = null;
+            }
+        }
+
+        public Transaction FallbackTransaction
+        {
+            get => fallbackTransaction;
+            set
+            {
+                fallbackTransaction = value;
+                hash = null;
+            }
+        }
+
+        public int Size => mainTransaction.Size + fallbackTransaction.Size + witness.Size;
+
+
+        public void DeserializeUnsigned(BinaryReader reader)
+        {
+            mainTransaction = reader.ReadSerializable<Transaction>();
+            fallbackTransaction = reader.ReadSerializable<Transaction>();
+        }
+
+        public void Deserialize(BinaryReader reader)
+        {
+            DeserializeUnsigned(reader);
+            witness = reader.ReadSerializable<Witness>();
+        }
+
+        public void Serialize(BinaryWriter writer)
+        {
+            SerializeUnsigned(writer);
+            writer.Write(witness);
+        }
+
+        public void SerializeUnsigned(BinaryWriter writer)
+        {
+            writer.Write(mainTransaction);
+            writer.Write(fallbackTransaction);
+        }
+
+        public UInt160[] GetScriptHashesForVerifying(DataCache snapshot)
+        {
+            return new UInt160[] { fallbackTransaction.Signers[1].Account };
+        }
+
+        public bool Verify(ProtocolSettings settings)
+        {
+            var nKeysMain = MainTransaction.GetAttributes<NotaryAssisted>();
+            if (!nKeysMain.Any()) return false;
+            if (nKeysMain.ToArray()[0].NKeys == 0) return false;
+            if (!fallbackTransaction.Script.SequenceEqual(FallbackFixedScript)) return false;
+            if (FallbackTransaction.Signers.Length != 2) return false;
```

Comment:
It'd be nice to have consistent code in this function wrt `fallbackTransaction`/`FallbackTransaction`.